src/System.Private.CoreLib/shared/System/Number.DiyFp.cs

```diff
@@ -8,121 +8,141 @@ namespace System
 
-                int lengthDelta = (SignificandLength - 53);
-                f <<= lengthDelta;
-                e -= lengthDelta;
+                ulong tmp = (bd >> 32) + (uint)(ad) + (uint)(bc);
 
-                result = new DiyFp((ulong)(f), e);
-            }
+                // By adding (1UL << 31) to tmp, we round the final result.
+                // Halfway cases will be rounded up.
 
-            public static void Minus(ref DiyFp lhs, ref DiyFp rhs, out DiyFp result)
-            {
-                result = lhs;
-                result.Minus(ref rhs);
+                tmp += (1U << 31);
+
+                return new DiyFp((ac + (ad >> 32) + (bc >> 32) + (tmp >> 32)), (e + other.e + SignificandSize));
             }
 
-            public static void Multiply(ref DiyFp lhs, ref DiyFp rhs, out DiyFp result)
+            public DiyFp Normalize()
             {
-                result = lhs;
-                result.Multiply(ref rhs);
+                // This method is mainly called for normalizing boundaries.
+                //
+                // We deviate from the reference implementation by just using
+                // our LeadingZeroCount function so that we only need to shift
+                // and subtract once.
+
+                Debug.Assert(f != 0);
+                int lzcnt = (int)(BigInteger.LeadingZeroCount(f));
+                return new DiyFp((f << lzcnt), (e - lzcnt));
             }
 
-            public void Minus(ref DiyFp rhs)
+            // The exponents of both numbers must be the same.
+            // The significand of this must be bigger than the significand of other.
+            // The result will not be normalized.
+            public DiyFp Subtract(in DiyFp other)
             {
-                Debug.Assert(_e == rhs._e);
-                Debug.Assert(_f >= rhs._f);
-
-                _f -= rhs._f;
+                Debug.Assert(e == other.e);
+                Debug.Assert(f >= other.f);
+                return new DiyFp((f - other.f), e);
             }
 
-            public void Multiply(ref DiyFp rhs)
+            private void GetBoundaries(int implicitBitIndex, out DiyFp mMinus, out DiyFp mPlus)
             {
-                ulong lf = _f;
-                ulong rf = rhs._f;
-
-                uint a = (uint)(lf >> 32);
-                uint b = (uint)(lf);
-
-                uint c = (uint)(rf >> 32);
-                uint d = (uint)(rf);
-
-                ulong ac = ((ulong)(a) * c);
-                ulong bc = ((ulong)(b) * c);
-                ulong ad = ((ulong)(a) * d);
-                ulong bd = ((ulong)(b) * d);
-
-                ulong tmp = (bd >> 32) + (uint)(ad) + (uint)(bc);
-                tmp += (1UL << 31);
+                mPlus = new DiyFp(((f << 1) + 1), (e - 1)).Normalize();
+
+                // The boundary is closer if the sigificand is of the form:
+                //      f == 2^p-1
+                //
+                // Think of v = 1000e10 and v- = 9999e9
```

Comment:
What's `v-`? Is it meant to be something like v-prime or does the negative sign after have some mathematical meaning?

Reply:
It is a mathematical symbol used in the paper. It means the lower boundary of `v`